### PR TITLE
Move self-test from NexusIO to test_Nexus

### PR DESCRIPTION
**Pull request by souravsingh**

### CODE REVIEW

File: Bio/AlignIO/NexusIO.py [hunk 1]
```diff
@@ -140,75 +140,4 @@ def _classify_alphabet_for_nexus(self, alphabet):
...
-    handle.seek(0)
-    print(handle.read())
-
-    handle = StringIO()
-    try:
-        NexusWriter(handle).write_file([a, a])
-        assert False, "Should have rejected more than one alignment!"
-    except ValueError:
-        pass
+    from Bio._utils import run_doctest
```
Comment (peterjc) [inline Bio/AlignIO/NexusIO.py]: This line doesn't do anything - you import the function ``run_doctest`` but don't use it:

``` python
from Bio._utils import run_doctest
```

File: Tests/test_Nexus.py [hunk 2]
```diff
@@ -438,7 +438,72 @@ def test_NexusComments(self):
                           (None, 0.20039426358269385, None, '[&height_95%_HPD={0.3110921040545068,0.38690865205576275},length_range={0.09675588357303178,0.4332959544380489},length_95%_HPD={0.16680375169879613,0.36500804261814374}]'),
                           ('9', 0.055354097721950546, None, '[&rate_range={1.3E-5,0.10958320752991428},height_95%_HPD={0.309132419999969,0.3091324199999691},length_range={3.513906814545109E-4,0.4381986285528381},height_median=0.309132419999969,length_95%_HPD={0.003011577063374571,0.08041621647998398}]'),
                           ('5', 0.055354097721950546, None, '[&rate_range={1.3E-5,0.10958320752991428},height_95%_HPD={0.309132419999969,0.3091324199999691},length_range={3.865051168833178E-5,0.4391594442572986},height_median=0.309132419999969,length_95%_HPD={0.003011577063374571,0.08041621647998398}]')])
+class TestSelf(unittest.TestCase):
+    def doctest(self):
+        print("Repeated names without a TAXA block")
+    handle = StringIO("""#NEXUS
```
Comment (peterjc) [inline Tests/test_Nexus.py]: The indentation is broken here

File: Tests/test_Nexus.py [hunk 3]
```diff
@@ -438,7 +438,72 @@ def test_NexusComments(self):
                           (None, 0.20039426358269385, None, '[&height_95%_HPD={0.3110921040545068,0.38690865205576275},length_range={0.09675588357303178,0.4332959544380489},length_95%_HPD={0.16680375169879613,0.36500804261814374}]'),
                           ('9', 0.055354097721950546, None, '[&rate_range={1.3E-5,0.10958320752991428},height_95%_HPD={0.309132419999969,0.3091324199999691},length_range={3.513906814545109E-4,0.4381986285528381},height_median=0.309132419999969,length_95%_HPD={0.003011577063374571,0.08041621647998398}]'),
                           ('5', 0.055354097721950546, None, '[&rate_range={1.3E-5,0.10958320752991428},height_95%_HPD={0.309132419999969,0.3091324199999691},length_range={3.865051168833178E-5,0.4391594442572986},height_median=0.309132419999969,length_95%_HPD={0.003011577063374571,0.08041621647998398}]')])
+class TestSelf(unittest.TestCase):
+    def doctest(self):
```
Comment (peterjc) [inline Tests/test_Nexus.py]: Using the Python ``unittest`` framework you would need to name the method something starting ``test_``, like ``test_repeated_names_no_taxa`` based on the description of the first test.

File: Tests/test_Nexus.py [hunk 4]
```diff
@@ -438,7 +438,72 @@ def test_NexusComments(self):
...
+    end; 
+    """)  # noqa for pep8 W291 trailing whitespace
+    for a in NexusIterator(handle):
+        print(a)
+        for r in a:
+            print("%r %s %s" % (r.seq, r.name, r.id))
+    print("Done")
 
+    print("")
+    print("Repeated names with a TAXA block")
```
Comment (peterjc) [inline Tests/test_Nexus.py]: This fragement from the original self-test would become a second method in the ``unittest`` framework, e.g. defined as a method named ``test_repeated_names_with_taxa``

File: Tests/test_Nexus.py [hunk 5]
```diff
@@ -438,7 +438,72 @@ def test_NexusComments(self):
...
+    ;
+    end; 
+    """)  # noqa for pep8 W291 trailing whitespace
+    for a in NexusIterator(handle):
+        print(a)
+        for r in a:
+            print("%r %s %s" % (r.seq, r.name, r.id))
+    print("Done")
+    print("")
+    print("Reading an empty file")
```
Comment (peterjc) [inline Tests/test_Nexus.py]: This would become a third ``unittest`` method, perhaps called ``test_empty_file``?

**Comment by peterjc** [comment]
Have you been able to run the tests locally yet? TravisCI is failing due to an indentation problem:

```
======================================================================
ERROR: test_Nexus
----------------------------------------------------------------------
Traceback (most recent call last):
  File "run_tests.py", line 420, in runTest
    suite = loader.loadTestsFromName(name)
  File "/opt/python/3.4.2/lib/python3.4/unittest/loader.py", line 105, in loadTestsFromName
    module = __import__('.'.join(parts_copy))
  File "/home/travis/build/biopython/biopython/Tests/test_Nexus.py", line 506
    pass
       ^
IndentationError: expected an indented block
----------------------------------------------------------------------
```

Have you tried this locally?

```
cd Tests
python test_Nexus.py
```

or:

```
cd Tests
python run_tests.py test_Nexus
```

**Comment by peterjc** [comment]
P.S. If you've not seen it, the ``unittest`` documentation is at https://docs.python.org/2/library/unittest.html (Python 2 version, its very similar for Python 3)

**Comment by souravsingh** [comment]
@peterjc Sorry for such a bad fix. I didn't quite understand about what the issue meant. I will take care of this from next time onwards.

**Comment by peterjc** [comment]
@souravsingh Don't worry too much - you're keen and learning more about Python (and that's a good outcome in itself), and with a little guidance I think you can get this change done :)

**Comment by souravsingh** [comment]
@peterjc The test failures in Travis seem unrelated to the Self test. How do I fix it?

**Comment by peterjc** [comment]
Travis reports this failure:

```
...
======================================================================
FAIL: test_NexusTest1 (test_Nexus.NexusTest1)
Test Nexus module
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_Nexus.py", line 332, in test_NexusTest1
    """)  # noqa for pep8 W291 trailing whitespace
AssertionError: 'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43     \n[C]    2.40      .       2.28     2.12     2.14     \n[G]    2.57     2.28      .       2.31     2.31     \n[T]    2.43     2.12     2.31      .       2.14     \n[-]    2.43     2.14     2.31     2.14      .       \n;\n' != 'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43\n[C]    2.40      .       2.28     2.12     2.14\n[G]    2.57     2.28      .       2.31     2.31\n[T]    2.43     2.12     2.31      .       2.14\n[-]    2.43     2.14     2.31     2.14      .\n;\n'
======================================================================
FAIL: test_TreeTest1 (test_Nexus.NexusTest1)
Test Tree module.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_Nexus.py", line 375, in test_TreeTest1
    self.assertEqual(len(output.split("\n")), len(expected.split("\n")))
AssertionError: 22 != 21
----------------------------------------------------------------------
Ran 263 tests in 537.893 seconds
FAILED (failures = 1)
```

i.e. ``test_NexusTest1`` files, because these values differ:

```
'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43     \n[C]    2.40      .       2.28     2.12     2.14     \n[G]    2.57     2.28      .       2.31     2.31     \n[T]    2.43     2.12     2.31      .       2.14     \n[-]    2.43     2.14     2.31     2.14      .       \n;\n'
```

and:

```
'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43\n[C]    2.40      .       2.28     2.12     2.14\n[G]    2.57     2.28      .       2.31     2.31\n[T]    2.43     2.12     2.31      .       2.14\n[-]    2.43     2.14     2.31     2.14      .\n;\n'
```

Note there is a difference in white space before some of the ``\n`` characters.

Also ``test_TreeTest1`` fails (22 lines instead of 21 lines).

What happens on your machine if you do:

```
python setup.py build
cd Tests
python run_tests.py test_Nexus.py
```

or:

```
python test_Nexus.py
```

**Comment by souravsingh** [comment]
@peterjc I ran test_Nexus in my local fork and it fails. Is the error due to the PEP8 fixes I pushed?

File: Tests/test_Nexus.py [hunk 6]
```diff
@@ -315,11 +323,11 @@ def test_NexusTest1(self):
         """\
 usertype matrix_test stepmatrix=5
         A        C        G        T        -
-[A]     .       2.40     2.57     2.43     2.43     
-[C]    2.40      .       2.28     2.12     2.14     
-[G]    2.57     2.28      .       2.31     2.31     
-[T]    2.43     2.12     2.31      .       2.14     
-[-]    2.43     2.14     2.31     2.14      .       
+[A]     .       2.40     2.57     2.43     2.43
```
Comment (peterjc) [inline Tests/test_Nexus.py]: I don't think you should have changed the whitespace here - this is where TravisCI reports the output does not match this given text:

```
======================================================================
FAIL: test_NexusTest1 (test_Nexus.NexusTest1)
Test Nexus module
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_Nexus.py", line 332, in test_NexusTest1
    """)  # noqa for pep8 W291 trailing whitespace
AssertionError: 'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43     \n[C]    2.40      .       2.28     2.12     2.14     \n[G]    2.57     2.28      .       2.31     2.31     \n[T]    2.43     2.12     2.31      .       2.14     \n[-]    2.43     2.14     2.31     2.14      .       \n;\n' != 'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43\n[C]    2.40      .       2.28     2.12     2.14\n[G]    2.57     2.28      .       2.31     2.31\n[T]    2.43     2.12     2.31      .       2.14\n[-]    2.43     2.14     2.31     2.14      .\n;\n'
======================================================================
FAIL: test_TreeTest1 (test_Nexus.NexusTest1)
Test Tree module.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_Nexus.py", line 375, in test_TreeTest1
    self.assertEqual(len(output.split("\n")), len(expected.split("\n")))
AssertionError: 22 != 21
----------------------------------------------------------------------
```

**Comment by peterjc** [comment]
Looking better, still some pep8 things:

```
$ pep8 --ignore E122,E123,E126,E127,E128,E241,E402,E501,E731 Tests/
Tests/test_Nexus.py:130:53: W291 trailing whitespace
Tests/test_Nexus.py:131:93: W291 trailing whitespace
Tests/test_Nexus.py:132:94: W291 trailing whitespace
Tests/test_Nexus.py:254:39: W291 trailing whitespace
Tests/test_Nexus.py:255:47: W291 trailing whitespace
Tests/test_Nexus.py:256:175: W291 trailing whitespace
Tests/test_Nexus.py:257:114: W291 trailing whitespace
Tests/test_Nexus.py:258:52: W291 trailing whitespace
Tests/test_Nexus.py:259:54: W291 trailing whitespace
Tests/test_Nexus.py:260:58: W291 trailing whitespace
Tests/test_Nexus.py:261:103: W291 trailing whitespace
Tests/test_Nexus.py:262:87: W291 trailing whitespace
Tests/test_Nexus.py:263:106: W291 trailing whitespace
Tests/test_Nexus.py:264:80: W291 trailing whitespace
Tests/test_Nexus.py:265:85: W291 trailing whitespace
Tests/test_Nexus.py:266:85: W291 trailing whitespace
Tests/test_Nexus.py:267:122: W291 trailing whitespace
Tests/test_Nexus.py:300:55: W291 trailing whitespace
Tests/test_Nexus.py:301:77: W291 trailing whitespace
Tests/test_Nexus.py:302:82: W291 trailing whitespace
Tests/test_Nexus.py:303:82: W291 trailing whitespace
Tests/test_Nexus.py:304:26: E124 closing bracket does not match visual indentation
ERROR: InvocationError: '/home/travis/build/biopython/biopython/.tox/pep8/bin/pep8 --ignore E122,E123,E126,E127,E128,E241,E402,E501,E731 Tests/'
```

and one test failing... looks nearly there though?

File: Tests/test_Nexus.py [hunk 7]
```diff
@@ -13,7 +13,6 @@
 import sys
 from Bio._py3k import StringIO
 from Bio._py3k import range
-from Bio.AlignIO import NexusIO
```
Comment (peterjc) [inline Tests/test_Nexus.py]: Based on the changes below, you probably meant to replace this with:

``` python
from Bio.AlignIO.NexusIO import NexusIterator
```

This explains at least some of the unit-test failures reported by TravisCI.

**Comment by souravsingh** [comment]
My Bad. I will patch it up

**Comment by peterjc** [comment]
You still have one test failure, as noted in earlier:

```
======================================================================
FAIL: test_NexusTest1 (test_Nexus.NexusTest1)
Test Nexus module
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_Nexus.py", line 324, in test_NexusTest1
    """)  # noqa for pep8 W291 trailing whitespace
AssertionError: 'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43     \n[C]    2.40      .       2.28     2.12     2.14     \n[G]    2.57     2.28      .       2.31     2.31     \n[T]    2.43     2.12     2.31      .       2.14     \n[-]    2.43     2.14     2.31     2.14      .       \n;\n' != 'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43\n[C]    2.40      .       2.28     2.12     2.14\n[G]    2.57     2.28      .       2.31     2.31\n[T]    2.43     2.12     2.31      .       2.14\n[-]    2.43     2.14     2.31     2.14      .\n;\n'
======================================================================
FAIL: test_TreeTest1 (test_Nexus.NexusTest1)
Test Tree module.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_Nexus.py", line 365, in test_TreeTest1
    self.assertEqual(len(output.split("\n")), len(expected.split("\n")))
AssertionError: 22 != 21
----------------------------------------------------------------------
```

This looks like a small white space change:

```
'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43     \n[C]    2.40      .       2.28     2.12     2.14     \n[G]    2.57     2.28      .       2.31     2.31     \n[T]    2.43     2.12     2.31      .       2.14     \n[-]    2.43     2.14     2.31     2.14      .       \n;\n'
```

vs:
```
'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43\n[C]    2.40      .       2.28     2.12     2.14\n[G]    2.57     2.28      .       2.31     2.31\n[T]    2.43     2.12     2.31      .       2.14\n[-]    2.43     2.14     2.31     2.14      .\n;\n'
```

Note the difference in spaces before some of the new lines.

**Comment by souravsingh** [comment]
@peterjc The modifications I made pass the tests. Did I do something wrong here, since the matrix look the same.

**Comment by peterjc** [comment]
According to TravisCI, the tests still fail https://travis-ci.org/biopython/biopython/builds/177298034 with what looks like a minor white space difference in the matrix output:

e.g. Python 2.7

```
======================================================================
FAIL: test_NexusTest1 (test_Nexus.NexusTest1)
Test Nexus module
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_Nexus.py", line 324, in test_NexusTest1
    """)  # noqa for pep8 W291 trailing whitespace
AssertionError: 'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43     \n[C]    2.40      .       2.28     2.12     2.14     \n[G]    2.57     2.28      .       2.31     2.31     \n[T]    2.43     2.12     2.31      .       2.14     \n[-]    2.43     2.14     2.31     2.14      .       \n;\n' != 'usertype matrix_test stepmatrix=5\n        A        C        G        T        -\n[A]     .       2.40     2.57     2.43     2.43\n[C]    2.40      .       2.28     2.12     2.14\n[G]    2.57     2.28      .       2.31     2.31\n[T]    2.43     2.12     2.31      .       2.14\n[-]    2.43     2.14     2.31     2.14      .  \n;\n'
----------------------------------------------------------------------
```

Or, Python 3.5,

```
======================================================================
FAIL: test_NexusTest1 (test_Nexus.NexusTest1)
Test Nexus module
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/biopython/biopython/Tests/test_Nexus.py", line 324, in test_NexusTest1
    """)  # noqa for pep8 W291 trailing whitespace
AssertionError: 'user[120 chars] 2.43     \n[C]    2.40      .       2.28     [181 chars]n;\n' != 'user[120 chars] 2.43\n[C]    2.40      .       2.28     2.12 [156 chars]n;\n'
Diff is 889 characters long. Set self.maxDiff to None to see it.
----------------------------------------------------------------------
```

(Ignore the TravisCI failure on PyPy3, there is currently a problem with dependencies there)

If this is working for you, there is something different about your setup. Are you running on Linux, Windows or Mac OS? Which version(s) of Python are you using?

**Comment by souravsingh** [comment]
I am using Linux( Ubuntu 16.04) and Python 2.7

**Comment by souravsingh** [comment]
@peterjc Is the patch ready for merging?

**Comment by peterjc** [comment]
Yay :+1: - working tests on TravisCI (except PyPy3 which is not your fault)

**Comment by peterjc** [comment]
Note I will squash this into a single commit when I merge it.

Are you happy to dual license your contributions under both the Biopython Licence Agreement AND the 3-Clause BSD license? See #898 

Are you happy to be thanked by name in the ``NEWS`` and ``CONTRIB`` files?

**Comment by souravsingh** [comment]
I give my permission to license my contributions under the BSD 3-clause license.

It would be great if my name is mentioned as part of contributors to Biopython. I wish to keep contributing to Biopython in future.

**Comment by peterjc** [comment]
Thank you - your changes are now included as commit https://github.com/biopython/biopython/commit/b7ad275a874dd2744bc193895924b0b5a94d5e0e with the acknowledgement in https://github.com/biopython/biopython/commit/a06363a82c460dea31d29faaab5102abd00695cb

**Pull request merged**